### PR TITLE
8 Format input data for model training (this time for GNNs)

### PR DESCRIPTION
**Pull request by cbarkr**
# This PR
- Is huge, sorry
- Updates the README with new info / fixes some issues
- Gets rid of / moves / and adds scripts (- `rip.py`, `scripts/make.py` -> `make.py, + `train.py`)
- Adds VexIR2Vec's ([paper](https://arxiv.org/abs/2312.00507), [source](https://github.com/IITH-Compilers/VexIR2Vec/tree/main)) model and relevant source code (model in `vole/models/vexir2vec.model`, source in `vole/utils/vexir2vec/*`)
- Adds a number of utils (`vole/utils/*`):
  - `cfg.py`: Basic CFG stuff (it was far more substantial half way through completing this task, but that functionality has since been replaced for the most part)
  - `embeddings.py`: Contains wrappers / adapters for interfacing with VexIR2Vec's spaghetti source code, including a number of rewrites / refactors of said spaghetti to make it more efficient and/or readable
  - `graph.py`: Some utils for working with networkx and torch_geometric graph data structures (like `cfg.py`, was quite substantial at one point, but has been refactored away to very little). Most notably, contains `normalize_edge_attributes`, which is **necessary** for adapting our CFGs to `torch_geometric.data.Data` using `torch_geometric.utils.convert.from_networkx` since it expects all edges to have the same attributes (which angr does not guarantee!)
  - `io.py`: Basically just houses a silly pathlib crawler. Had bigger plans for this, but they haven't come to fruition (yet?)
  - `train.py`: Basically just houses a silly util for splitting up corpus data. Again, had bigger plans for this that haven't seen the light of day (or the dark of night)
- Most notably, adds `train.py`, the _pièce de résistance_ of this PR, which combines all of our progress thus far to, in this order:
1. Receive a CWE-ID and path to compiled training data as input
2. Recursively crawl through the path and identify binaries matching the CWE-ID
3. Split and shuffle these matching binaries according to our predetermined training split
4. For each path in the split:
    1. Generate a CFG
    2. For each function in the CFG (note: since the SARD test cases have no `main`, these will be the functions we want to train on):
        1. Generate a subgraph / subCFG of the function
        2. Perform a random walk over the function's blocks (note: the walk is quite long and is likely to span the entirety of the blocks for our test small cases)
        3. Collect the IR for each block, normalize and perform some preprocessing on it, generate a vector embedding of it, and finally add it to a function-level vector embedding
        4. Run the VexIR2Vec model on said IR (and other data) to get a singular vector embedding for each function
        5. Insert a "good" (0) or  "bad" (1) label and the vector embedding to each node in the function's subgraph
        6. Convert this subgraph into a `torch_geometric.data.Data` using `torch_geometric.utils.convert.from_networkx`
5. Output a list of such data to be fed into a GCN

# Notes
- By using VexIR2Vec, we are making a bit of a tradeoff of fidelity/granularity with accuracy
- Ideally, we want block- (or even better, statement-) level IR vector embeddings as these would localize the vulnerabilities quite precisely
- However, VexIR2Vec operates on function-level embeddings
- That being said, VexIR2Vec is likely to be WAY more accurate than any model we could train ourselves
- Thus, I think it's a fair tradeoff (and why I have opted to refactor my implementation to function-level embeddings (which are stored for each node in the function))

# Future Work
- Further refactor the README to remove PyPy/Rust altogether
- Change GCC version in Dockerfile since it's no longer necessary to be on the bleeding edge
- Figure out a way to load the model with `weights_only=True` to mitigate the risk of RCE
- Figure out a way to apply more granular labelling (currently nodes are labelled as good or bad based on their parent - but with the current VexIR2Vec implementation, this might be a no-op)
- Identify an ideal normalization level for the `EmbeddingsWrapper` - [the docs](https://github.com/IITH-Compilers/VexIR2Vec/tree/main/embeddings/pre-training#data-generation) say "unless required, choose `3`" but the [source code](https://github.com/IITH-Compilers/VexIR2Vec/blob/43538167644db81cbfda89716c113b483aa9fd06/embeddings/pre-training/driver.py#L297) uses `1` by default
- Train the damn GCN!

# Testing This PR
1. Follow the instructions in https://fasttext.cc/docs/en/crawl-vectors.html to get the pretrained FastText model and reduce it to a dimensionality of 100
2. Move the resulting file `cc.en.100.bin` to `vole/models/`
3. Follow the instructions in the README.md to setup your development environment, acquire the SARD test data, and compile some SARD test cases
4. Nothing will really happen with `train.py` as of right now, so perhaps insert a `print` statement or two here and there to see what kinda data we're working with
5. For example, in `vole/train.py`:

```diff
def prepare_data_for_split(
    split: list[pathlib.Path], ir_embed: IREmbeddings
) -> list:
    split_data = []

    for path in split:
        proj, cfg = get_project_cfg(path)
        embeddings = ir_embed.get_function_embeddings(proj, cfg)
+       print(embeddings)
```

6. Or `vole/utils/embeddings.py`:

```diff
class IREmbeddings:
    def __init__(self):
        self.vocab = pathlib.Path(PARENT / "./vexir2vec/vocabulary.txt")
        self.model = pathlib.Path(PARENT / "../models/vexir2vec.model")
        self.embeddings = EmbeddingsWrapper(self.vocab)

        # Patch the resolution of the model's source at runtime
        sys.modules["model_OTA"] = utils.vexir2vec.model_OTA

        # TODO: Figure out if it's possible to load the model with weights_only=True
        self.vexir2vec = torch.load(
            self.model, map_location=DEVICE, weights_only=False
        )
        self.vexir2vec.eval()
+       print(self.vexir2vec)

# ...

#     def get_function_embeddings(
#          ...

           ir_vec = []
           for data in dataloader:
                with torch.no_grad():
                    opc_emb = data[0].to(DEVICE)
                    ty_emb = data[1].to(DEVICE)
                    arg_emb = data[2].to(DEVICE)
                    str_emb = data[3].to(DEVICE)
                    lib_emb = data[4].to(DEVICE)

                    res = self.vexir2vec(
                        opc_emb, ty_emb, arg_emb, str_emb, lib_emb
                    )
+                   print(res)
                    ir_vec.extend(res)
```

### CODE REVIEW

**Review by shabbiryusufali — APPROVED** [review]
LGTM (Provided everything works)

**Review by acey03 — COMMENTED** [review]
tldr (jk, works on my end)
